### PR TITLE
Add enrichment operation

### LINKED ISSUE #49: Add Enrichment Analysis
Enrichment Analysis is a method currently used by Juan. His code can be found here: 
https://github.com/broadinstitute/DeepProfilerExperiments/blob/master/ta-orf/04-downstream-analysis.ipynb

I would like to add the enrichment operation to the list of operations while using the given tools and adapting the code from Juan.

### PR DESCRIPTION
**Pull request by michaelbornholdt**
fixes #49

### CODE REVIEW

**Comment by michaelbornholdt** [comment]
@gwaygenomics Here is my PR 

## What is the nature of your change?

- [X] Enhancement (adds functionality).

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [ ] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have deleted all non-relevant text in this pull request template.

**Comment by gwaygenomics** [comment]
nice! Looking good. What would you like me to look at? A quick first pass makes me think you still need to add some tests. See https://github.com/cytomining/cytominer-eval/tree/master/cytominer_eval/tests/test_operations for some examples

**Comment by michaelbornholdt** [comment]
Hey @gwaygenomics, 
All the checks are passing - that's good :)

**Comment by michaelbornholdt** [comment]
Whats left to do?

**Comment by gwaygenomics** [comment]
Now i'll review! (this is less daunting than it may sound) :)

**Comment by michaelbornholdt** [comment]
Btw, I cant set assignees and co right- because I am missing the right?

**Comment by gwaygenomics** [comment]
I think so. you did right to poke with a direct mention, i can self-assign

File: cytominer_eval/evaluate.py [hunk 1]
```diff
@@ -16,6 +16,9 @@
     mp_value,
 )
 
+# no idea why I need to do this, but otherwise it would give the module object error
+from cytominer_eval.operations.enrichment import enrichment
```
Comment (gwaygenomics) [inline cytominer_eval/evaluate.py:20]: Please add to [`cytominer_eval/operations/__init__.py`](https://github.com/cytomining/cytominer-eval/blob/master/cytominer_eval/operations/__init__.py):

```python
from .enrichment import enrichment
```

Then it should work just fine

File: cytominer_eval/evaluate.py [hunk 2]
```diff
@@ -30,6 +33,7 @@ def evaluate(
     grit_control_perts: List[str] = ["None"],
     grit_replicate_summary_method: str = "mean",
     mp_value_params: dict = {},
+    percentile: float = 0.5,
```
Comment (gwaygenomics) [inline cytominer_eval/evaluate.py:36]:
```suggestion
    enrichment_percentile: float = 0.5,
```

Just to make sure it's clear that `percentile` is only operational with `operation="enrichment"`

File: cytominer_eval/operations/enrichment.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
+"""Function to calculate the enrichtment score for a given similarity matrix.
```
Comment (gwaygenomics) [inline cytominer_eval/operations/enrichment.py:1]:
```suggestion
"""Function to calculate the enrichment score for a given similarity matrix.
```

to fix typo

File: cytominer_eval/operations/enrichment.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    set_grit_column_info,
+    assert_melt,
+)
+
+
+def enrichment(
+    similarity_melted_df: pd.DataFrame, replicate_groups: List[str], percentile: float,
+) -> dict:
+    """Calculate the enrichment score. This score is based on the fisher exact odds score. Similar to the other functions, the closest connections are determined and checked with the replicates.
+    This score effectively calculates how much better the distribution of correct connections is from a random sample.
```
Comment (gwaygenomics) [inline cytominer_eval/operations/enrichment.py:20]:
```suggestion
    This score effectively calculates how much better the distribution of correct connections is compared to random.
```

File: cytominer_eval/operations/enrichment.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+from .util import assign_replicates, calculate_grit, check_grit_replicate_summary_method
+from cytominer_eval.transform.util import (
+    set_pair_ids,
+    set_grit_column_info,
+    assert_melt,
+)
+
+
+def enrichment(
+    similarity_melted_df: pd.DataFrame, replicate_groups: List[str], percentile: float,
```
Comment (gwaygenomics) [inline cytominer_eval/operations/enrichment.py:17]: Just checking, is it intentional that `percentile` doesn't specify a default?

File: cytominer_eval/operations/enrichment.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    replicate_groups : List
+        a list of metadata column names in the original profile dataframe to use as
+        replicate columns.
+    percentile :  float
+        Determines what percentage of top connections used for the enrichment calculation.
+
+    Returns
+    -------
+    pandas.DataFrame
+        percentile, threshold, ods ration and p value
```
Comment (gwaygenomics) [inline cytominer_eval/operations/enrichment.py:37]:
```suggestion
        percentile, threshold, odds ratio and p value
```

File: cytominer_eval/operations/enrichment.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        :py:func:`cytominer_eval.transform.transform.metric_melt`.
+    replicate_groups : List
+        a list of metadata column names in the original profile dataframe to use as
+        replicate columns.
+    percentile :  float
+        Determines what percentage of top connections used for the enrichment calculation.
+
+    Returns
+    -------
+    pandas.DataFrame
```
Comment (gwaygenomics) [inline cytominer_eval/operations/enrichment.py:36]:
```suggestion
    dict
```

Is this true?

File: cytominer_eval/operations/enrichment.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        )
+    )
+    v22 = len(
+        replicate_truth_df.query(
+            "group_replicate==False and similarity_metric<=@threshold"
+        )
+    )
+
+    V = np.asarray([[v11, v12], [v21, v22]])
+    # print(percentile, threshold)
```
Comment (gwaygenomics) [inline cytominer_eval/operations/enrichment.py:68]: Can you cleanup print debug statements?

File: cytominer_eval/operations/enrichment.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    # print(V, np.sum(V))
+    r = scipy.stats.fisher_exact(V, alternative="greater")
+    result = {
+        "percentile": percentile,
+        "threshold": threshold,
+        "ods_ratio": r[0],
+        "p-value": r[1],
+    }
+    # df = pd.DataFrame(data=d)
```
Comment (gwaygenomics) [inline cytominer_eval/operations/enrichment.py:77]: Should this function return a data frame? if not, please delete

File: cytominer_eval/tests/test_operations/test_enrichment.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+replicate_groups = ["Metadata_broad_sample"]
+
+similarity_melted_df = metric_melt(
+    df=df,
+    features=features,
+    metadata_features=meta_features,
+    similarity_metric="pearson",
+    eval_metric="enrichment",
+)
+def test_enrichment():
```
Comment (gwaygenomics) [inline cytominer_eval/tests/test_operations/test_enrichment.py:41]: make sure to run `black` on the test files too. (there should be two new lines before a function definition)

File: demos/CellPainting_Demo.ipynb [hunk 11]
```diff
@@ -17,7 +17,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 1,
+   "execution_count": 36,
```
Comment (gwaygenomics) [inline demos/CellPainting_Demo.ipynb:20]: It looks like you're adding `enrichment` to the demo - that is great!

Before merging the cell painting demo, can you make sure to `restart kernel -> run all`. It is best data science practice to make sure the cells of a jupyter notebook are all executed sequentially.

File: cytominer_eval/evaluate.py [hunk 12]
```diff
@@ -99,6 +103,9 @@ def evaluate(
         Only used when `operation='mp_value'`. A key, item pair of optional parameters
         for calculating mp value. See also
         :py:func:`cytominer_eval.operations.util.default_mp_value_parameters`
+    percentile : float, optional
+        Only used when `operation='enrichment'`. Determines what percentage of top connections
```
Comment (gwaygenomics) [inline cytominer_eval/evaluate.py:107]:
```suggestion
        Only used when `operation='enrichment'`. Determines the percentage of top connections
```

File: cytominer_eval/tests/test_operations/test_enrichment.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            percentile=p,
+        )
+        result.append(r)
+    result_df = pd.DataFrame(result)
+
+    # check for correct shape and starts with 1.0
+    assert result_df.shape == (7, 4)
+    assert result_df.percentile[0] == 1.0
+    # check if the higher percentiles are larger than the small one
+    assert result_df.percentile[1] > result_df.percentile.iloc[-1]
```
Comment (gwaygenomics) [inline cytominer_eval/tests/test_operations/test_enrichment.py:56]: can you also add a test using `enrichment` through the `evaluate()` operation? We want to make sure that this connection is strong. (i.e. can you check that the result of `enrichment(percentile=p)` is equivalent to `evaluate(operation="enrichment", enrichment_percentile=p)`

**Review by gwaygenomics — CHANGES_REQUESTED** [review]
Looking very good. Thanks for this major contribution!

I make a couple of minor suggestions and add a few discussion points.

**Comment by gwaygenomics** [comment]
@michaelbornholdt - ready for you to take a look and address the PR comments

File: cytominer_eval/evaluate.py (hunk 1 above)
Comment (michaelbornholdt) [reply cytominer_eval/evaluate.py:20]: Done

File: cytominer_eval/evaluate.py (hunk 2 above)
Comment (michaelbornholdt) [reply cytominer_eval/evaluate.py:36]: Done

File: cytominer_eval/operations/enrichment.py (hunk 5 above)
Comment (michaelbornholdt) [reply cytominer_eval/operations/enrichment.py:17]: set it to 0.9

File: cytominer_eval/operations/enrichment.py (hunk 8 above)
Comment (michaelbornholdt) [reply cytominer_eval/operations/enrichment.py:68]: Done

File: cytominer_eval/operations/enrichment.py (hunk 9 above)
Comment (michaelbornholdt) [reply cytominer_eval/operations/enrichment.py]: Done

File: cytominer_eval/tests/test_operations/test_enrichment.py (hunk 10 above)
Comment (michaelbornholdt) [reply cytominer_eval/tests/test_operations/test_enrichment.py:41]: Done

File: demos/CellPainting_Demo.ipynb (hunk 11 above)
Comment (michaelbornholdt) [reply demos/CellPainting_Demo.ipynb:20]: ok, will do

File: cytominer_eval/tests/test_operations/test_enrichment.py (hunk 13 above)
Comment (michaelbornholdt) [reply cytominer_eval/tests/test_operations/test_enrichment.py:56]: Done

**Comment by michaelbornholdt** [comment]
Do the checks start running when all issues are resolved?

**Comment by michaelbornholdt** [comment]
@gwaygenomics Ready for review again. Notebook is currently running

**Review by gwaygenomics — COMMENTED** [review]
Very nice. One last comment, can you confirm that your ran the demo notebook within the provided demo environment as specified in the docstring [here](https://github.com/cytomining/cytominer-eval/blob/master/demos/CellPainting_Demo.ipynb)?

After you confirm this, then I will merge

**Comment by michaelbornholdt** [comment]
Yea thats correct

**Pull request merged**